### PR TITLE
[MI-2109]: Revamped UI and minor fixes

### PR DESCRIPTION
**Pull request by avas27JTG**
Revamped UI and minor fixes

### CODE REVIEW

File: webapp/src/components/card/subscription/styles.scss [hunk 1]
```diff
@@ -6,3 +6,11 @@
...
     }
 }
+
+.icon-label {
+    font-size: 16px;
+}
+
+.icon-mm {
+    margin: 0 3px 0 0;
+}
```
Comment (ayusht2810) [inline webapp/src/components/card/subscription/styles.scss:16]: add eof

File: webapp/src/containers/Rhs/projectDetails/index.tsx [hunk 2]
```diff
@@ -137,14 +147,16 @@ const ProjectDetails = (projectDetails: ProjectDetails) => {
...
+    }, [getWebsocketEventState(state).isSubscriptionDeleted]);
+
     return (
         <>
-            <BackButton onClick={handleResetProjectDetails}/>
-            <ToggleSwitch
-                active={showAllSubscriptions}
-                onChange={setShowAllSubscriptions}
-                label={plugin_constants.common.ToggleLabel}
-            />
```
Comment (ayusht2810) [inline webapp/src/containers/Rhs/projectDetails/index.tsx:147]: this part is updated in master

Comment (ayusht2810) [reply webapp/src/containers/Rhs/projectDetails/index.tsx:147]: many changes are not in sync with the master. please check if this is correct

File: webapp/src/containers/Rhs/projectList/index.tsx [hunk 3]
```diff
@@ -78,13 +83,13 @@ const ProjectList = () => {
                     title='Confirm Project Unlink'
                 />
             }
-            {isLoading && <LinearLoader/> }
+            {isLoading && <LinearLoader/>}
             {
                 isSuccess && (
-                    projectsList?.length > 0 ?
+                    sortedProjectList && sortedProjectList.length > 0 ?
```
Comment (ayusht2810) [inline webapp/src/containers/Rhs/projectList/index.tsx:89]: I think we don't need the first condition `sortedProjectList`.

File: webapp/src/containers/Rhs/projectList/index.tsx [hunk 4]
```diff
@@ -78,13 +83,13 @@ const ProjectList = () => {
...
                         <>
                             {
-                                projectsList.map((item: ProjectDetails) => (
+                                sortedProjectList?.map((item: ProjectDetails) => (
```
Comment (ayusht2810) [inline webapp/src/containers/Rhs/projectList/index.tsx:92]:
```suggestion
                                sortedProjectList.map((item: ProjectDetails) => (
```

File: webapp/src/containers/modals/LinkModal/index.tsx [hunk 5]
```diff
@@ -70,7 +70,7 @@ const LinkModal = () => {
             title='Link New Project'
             onHide={resetModalState}
             onConfirm={onConfirm}
-            confirmBtnText='Link new project'
+            confirmBtnText='Link New Project'
```
Comment (ayusht2810) [inline webapp/src/containers/modals/LinkModal/index.tsx:73]: why we have removed `ResultPanel` from this modal?

File: webapp/src/types/common/form.d.ts [hunk 6]
```diff
@@ -2,9 +2,10 @@ type FieldType = 'dropdown' | 'text' | 'timestamp'
 type ValidationTypes = 'isRequired' | 'maxCharLen' | 'minCharLen' | 'regex' | 'regexErrorMessage'
 type ErrorComponents = 'LinkProjectModal' | 'CreateTaskModal' | 'SubscribeModal'
 
-type SubscriptionModalFields = 'organization' | 'project' | 'eventType' | 'channelID' | 'timestamp'
-type CreateTaskModalFields = 'organization' | 'project' | 'type' | 'title' | 'description' | 'areaPath' | 'timestamp';
 type LinkProjectModalFields = 'organization' | 'project' | 'timestamp'
+type CreateTaskModalFields = 'organization' | 'project' | 'type' | 'title' | 'description' | 'areaPath' | 'timestamp'
+type SubscriptionModalFields = 'organization' | 'project' | 'eventType' | 'channelID' | 'timestamp'
+type ErrorComponents = 'SubscribeModal' | 'CreateTaskModal'
```
Comment (ayusht2810) [inline webapp/src/types/common/form.d.ts:8]: it is present twice

File: webapp/src/utils/index.ts [hunk 7]
```diff
@@ -6,7 +6,7 @@ import Constants from 'plugin_constants';
 import {getOrganizationList, getProjectList} from './filterData';
 import getErrorMessage from './errorHandling';
 
-const getBaseUrls = (): {pluginApiBaseUrl: string; mattermostApiBaseUrl: string} => {
+const getBaseUrls = (): { pluginApiBaseUrl: string; mattermostApiBaseUrl: string } => {
```
Comment (ayusht2810) [inline webapp/src/utils/index.ts:9]:
```suggestion
const getBaseUrls = (): {pluginApiBaseUrl: string; mattermostApiBaseUrl: string} => {
```

File: webapp/src/components/card/subscription/index.tsx [hunk 8]
```diff
@@ -5,27 +5,33 @@ import IconButton from 'components/buttons/iconButton';
...
+                        >
+                            {plugin_constants.SVGIcons.workEvent}
+                        </SVGWrapper>
+                    }
+                    labelExtraClassName='margin-left-5'
+                    value={plugin_constants.common.boardsEventTypeMap[eventType as EventType]}
                 />
                 <LabelValuePair
-                    label='Channel'
+                    labelIconClassName={`icon ${channelType === 'P' ? 'icon-lock-outline' : 'icon-globe'} icon-label`}
```
Comment (sooraj-shukla) [inline webapp/src/components/card/subscription/index.tsx:34]: Add `channelType` in constants.

File: webapp/src/components/card/subscription/styles.scss (hunk 1 above)
Comment (sooraj-shukla) [inline webapp/src/components/card/subscription/styles.scss:16]: No newline at the end of file

File: webapp/src/utils/filterData.ts [hunk 9]
```diff
@@ -19,4 +19,9 @@ export const getProjectList = (data: ProjectDetails[]) => {
     return projectList;
 };
 
-export const getCurrentChannelSubscriptions = (data: SubscriptionDetails[], channelID: string): SubscriptionDetails[] => data.filter(((subscription) => subscription.channelID === channelID));
```
Comment (ayusht2810) [inline webapp/src/utils/filterData.ts:22]: sync with master

File: webapp/src/components/toggleSwitch/index.tsx [hunk 10]
```diff
@@ -15,10 +14,10 @@ const ToggleSwitch = ({
     active,
     onChange,
     label,
-    labelPositioning = ToggleSwitchLabelPositioning.Left,
```
Comment (sooraj-shukla) [inline webapp/src/components/toggleSwitch/index.tsx:18]: Why this change?

File: webapp/src/components/toggleSwitch/index.tsx [hunk 11]
```diff
@@ -15,10 +14,10 @@ const ToggleSwitch = ({
...
+    labelPositioning = 'left',
 }: ToggleSwitchProps): JSX.Element => (
-    <div className={`toggle-switch-container d-flex align-items-center ${labelPositioning === ToggleSwitchLabelPositioning.Right && 'flex-row-reverse'}`}>
-        {label && <span className={labelPositioning === ToggleSwitchLabelPositioning.Left ? 'toggle-switch-label--left' : 'toggle-switch-label--right'}>{label}</span>}
+    <div className={`toggle-switch-container d-flex align-items-center ${labelPositioning === 'right' && 'flex-row-reverse justify-content-end'}`}>
+        {label && <span className={labelPositioning === 'left' ? 'toggle-switch-label--left' : 'toggle-switch-label--right'}>{label}</span>}
```
Comment (sooraj-shukla) [inline webapp/src/components/toggleSwitch/index.tsx:20]: Shouldn't we be using constants here for adding default positioning?

File: webapp/src/styles/_components.scss [hunk 12]
```diff
@@ -1,5 +1,5 @@
 .height-rhs {
-    height: 100%;
+    height: calc(100% - 150px); // 100% is the entire height and 150px is the space occupied by the button fixed at the bottom
```
Comment (sooraj-shukla) [inline webapp/src/styles/_components.scss:2]: It'll be good to add `150px` to variables.

File: webapp/src/styles/_utils.scss [hunk 13]
```diff
@@ -34,6 +50,19 @@
     display: flex;
 }
 
+.align-item-center {
```
Comment (sooraj-shukla) [inline webapp/src/styles/_utils.scss:53]: There is a util class available in mattermost for this.

File: webapp/src/styles/_utils.scss [hunk 14]
```diff
@@ -53,6 +82,10 @@
     }
 }
 
+.justify-content-end {
+    justify-content: flex-end;
```
Comment (sooraj-shukla) [inline webapp/src/styles/_utils.scss:86]: Same as above.

File: webapp/src/styles/_utils.scss [hunk 15]
```diff
@@ -34,6 +50,19 @@
     display: flex;
```
Comment (sooraj-shukla) [inline webapp/src/styles/_utils.scss:50]: There is a util class available in mattermost for this.

File: webapp/src/styles/_utils.scss [hunk 16]
```diff
@@ -93,3 +121,9 @@
 .overflow-auto {
     overflow: auto;
 }
+
+.text-truncate {
```
Comment (sooraj-shukla) [inline webapp/src/styles/_utils.scss:125]: `text-ellipses` class is used in a few places but that util class isn't there in the util styles.

File: webapp/src/containers/Rhs/projectDetails/index.tsx (hunk 2 above)
Comment (avas27JTG) [reply webapp/src/containers/Rhs/projectDetails/index.tsx:147]: okay will check when syncing with master

File: webapp/src/utils/index.ts [hunk 17]
```diff
@@ -58,14 +58,24 @@ export const getCreateTaskModalCommandArgs = (arr: Array<string>): CreateTaskCom
...
+export const sortProjectList = (project1: ProjectDetails, project2: ProjectDetails) => {
+    if (project1.projectName.toLocaleLowerCase() < project2.projectName.toLocaleLowerCase()) {
+        return -1;
+    }
+    if (project1.projectName.toLocaleLowerCase() > project2.projectName.toLocaleLowerCase()) {
+        return 1;
+    }
+    return 0;
+};
+
```
Comment (sooraj-shukla) [inline webapp/src/utils/index.ts:78]:
```suggestion
export const sortProjectList = (project1: ProjectDetails, project2: ProjectDetails) => project1.projectName.toLocaleLowerCase().localeCompare(project2.projectName.toLocaleLowerCase());

```

**Review by sooraj-shukla — CHANGES_REQUESTED** [review]

File: webapp/src/components/toggleSwitch/index.tsx (hunk 10 above)
Comment (avas27JTG) [reply webapp/src/components/toggleSwitch/index.tsx:18]: we should use type, not enum as using `type` is easier and bounds the values

File: webapp/src/components/toggleSwitch/index.tsx (hunk 11 above)
Comment (avas27JTG) [reply webapp/src/components/toggleSwitch/index.tsx:20]: we can use utilize TS type for this, using constant will lead to always importing that constant while passing as prop whereas type doesn't

File: webapp/src/styles/_utils.scss (hunk 13 above)
Comment (avas27JTG) [reply webapp/src/styles/_utils.scss:53]: okay but this might be used at many places so let's not depend on MM utils classes as there might be then utils for other things as well which we don't know

File: server/plugin/command.go [hunk 18]
```diff
@@ -61,9 +62,15 @@ func (p *Plugin) getAutoCompleteData() *model.AutocompleteData {
 	create := model.NewAutocompleteData("boards create [title] [description]", "", "Create a new task")
 	azureDevops.AddCommand(create)
 
-	subscribe := model.NewAutocompleteData("subscribe", "", "Add a subscription")
+	subscribe := model.NewAutocompleteData("boards subscribe", "", "Add a subscription")
```
Comment (ayusht2810) [inline server/plugin/command.go:65]:
```suggestion
	subscribe := model.NewAutocompleteData("boards subscribe", "", "Add a board subscription")
```

Comment (ayusht2810) [reply server/plugin/command.go:65]: maybe? to match with unsubscribe command

File: webapp/src/components/loader/linear/index.tsx [hunk 19]
```diff
@@ -2,8 +2,8 @@ import React from 'react';
 
 import './styles.scss';
 
-const LinearLoader = () : JSX.Element => (
-    <div className='linear-loader'>
+const LinearLoader = ({extraClass}: {extraClass?: string}) : JSX.Element => (
+    <div className={`linear-loader ${extraClass}`}>
```
Comment (ayusht2810) [inline webapp/src/components/loader/linear/index.tsx:6]: will this class contain undefined if we don't send this prop?

File: webapp/src/containers/modals/LinkModal/index.tsx [hunk 20]
```diff
@@ -30,11 +31,19 @@ const LinkModal = () => {
...
     const resetModalState = (isActionDone?: boolean) => {
         dispatch(toggleShowLinkModal({isVisible: false, commandArgs: [], isActionDone}));
         resetFormFields();
+        setShowResultPanel(false);
+    };
+
+    // Opens link project modal
+    const handleOpenLinkProjectModal = () => {
+        resetModalState();
+        dispatch(toggleShowLinkModal({isVisible: true, commandArgs: []}));
```
Comment (ayusht2810) [inline webapp/src/containers/modals/LinkModal/index.tsx:46]: we should dispatch first?

File: webapp/src/reducers/index.ts [hunk 21]
```diff
@@ -17,7 +17,7 @@ const reducers = combineReducers({
     createTaskModalSlice,
     subscriptionModalSlice,
     projectDetailsSlice,
-    userConnectedSlice,
```
Comment (ayusht2810) [inline webapp/src/reducers/index.ts:20]: shouldn't we remove this reducer as well `userConnectedSlice` if not used anywhere?

File: webapp/src/utils/filterData.ts (hunk 9 above)
Comment (ayusht2810) [reply webapp/src/utils/filterData.ts:22]: this is left to be synced.

File: webapp/src/containers/Rhs/projectDetails/index.tsx (hunk 2 above)
Comment (ayusht2810) [reply webapp/src/containers/Rhs/projectDetails/index.tsx:147]: this is left to be synced.

File: webapp/src/components/toggleSwitch/index.tsx (hunk 10 above)
Comment (sooraj-shukla) [reply webapp/src/components/toggleSwitch/index.tsx:18]: Here we're providing value using `enum/constant` instead of hard-coded values. Shouldn't we use `enum`/`constants` in such cases?

File: webapp/src/styles/_utils.scss (hunk 13 above)
Comment (sooraj-shukla) [reply webapp/src/styles/_utils.scss:53]: We can see all the utils in `mattermost-webapp`. And these are the things that are provided by `react-bootstrap` also, so we can depend on `react-bootstrap` if not on mm.

File: webapp/src/components/toggleSwitch/index.tsx (hunk 11 above)
Comment (sooraj-shukla) [reply webapp/src/components/toggleSwitch/index.tsx:20]: I guess while passing the props it'll be okay to use a string literal. But here in the component, we should use constant because this is something that's not expected to change.

File: webapp/src/containers/Rhs/projectDetails/index.tsx [hunk 22]
```diff
@@ -109,21 +121,20 @@ const ProjectDetails = (projectDetails: ProjectDetails) => {
...
             if (showAllSubscriptions) {
-                setSubscriptionList(subscriptionData);
+                setSubscriptionList(subscriptionListReturnedByApi);
             } else {
-                setSubscriptionList(getCurrentChannelSubscriptions(subscriptionData, currentChannelId));
+                setSubscriptionList(getCurrentChannelSubscriptions(subscriptionListReturnedByApi, currentChannelId));
             }
         }
-    }, [subscriptionData, showAllSubscriptions]);
+    }, [subscriptionListReturnedByApi]);
```
Comment (sooraj-shukla) [inline webapp/src/containers/Rhs/projectDetails/index.tsx:132]: Please check if this `useEffect` is running on each render. 
(It should be running because `subscriptionListReturnedByApi` is created again and again in each render cycle. In such cases, we need to memoize the value otherwise the dependency array won't have any performance improvement.)

File: webapp/src/containers/Rhs/projectDetails/index.tsx [hunk 23]
```diff
@@ -163,22 +176,26 @@ const ProjectDetails = (projectDetails: ProjectDetails) => {
                 description='Are you sure you want to delete this subscription?'
                 title='Confirm Delete Subscription'
             />
-            {isLoading && <LinearLoader/>}
-            <div className='d-flex'>
+            {isLoading && <LinearLoader extraClass='top-0'/>}
```
Comment (sooraj-shukla) [inline webapp/src/containers/Rhs/projectDetails/index.tsx:179]: `top-0` doesn't specify the meaning of what this class is used for. Can we name it so that it is more descriptive?

File: webapp/src/containers/Rhs/projectDetails/index.tsx [hunk 24]
```diff
@@ -163,22 +176,26 @@ const ProjectDetails = (projectDetails: ProjectDetails) => {
...
+            <ToggleSwitch
+                active={showAllSubscriptions}
+                onChange={handleToggle}
+                label={'Show All Subscriptions'}
```
Comment (sooraj-shukla) [inline webapp/src/containers/Rhs/projectDetails/index.tsx:183]: Shouldn't we use constant here?

File: webapp/src/containers/Rhs/projectList/index.tsx [hunk 25]
```diff
@@ -63,10 +66,11 @@ const ProjectList = () => {
 
     const {data, isSuccess, isLoading} = getApiState(plugin_constants.pluginApiServiceConfigs.getAllLinkedProjectsList.apiServiceName);
     const projectsList = data as ProjectDetails[];
+    const sortedProjectList = [...projectsList].sort(sortProjectList);
```
Comment (sooraj-shukla) [inline webapp/src/containers/Rhs/projectList/index.tsx:69]: Is `data` always an array? I think, `data` can be `undefined` as well. And in such case, it'll break.

File: webapp/src/containers/Rhs/projectList/index.tsx [hunk 26]
```diff
@@ -78,13 +82,13 @@ const ProjectList = () => {
...
-            {isLoading && <LinearLoader/> }
+            {isLoading && <LinearLoader/>}
             {
                 isSuccess && (
-                    projectsList?.length > 0 ?
+                    sortedProjectList.length > 0 ?
                         <>
                             {
-                                projectsList.map((item: ProjectDetails) => (
+                                sortedProjectList.map((item: ProjectDetails) => (
```
Comment (sooraj-shukla) [inline webapp/src/containers/Rhs/projectList/index.tsx:91]: We can improve the performance by memoizing the `sortedProjectList`. For now, it'll be re-created on each render and because of that, the whole list of projects will be re-rendered on each render no matter whether the projects are changed or not.

File: webapp/src/containers/Rhs/projectList/index.tsx [hunk 27]
```diff
@@ -99,14 +103,14 @@ const ProjectList = () => {
...
                                     className='plugin-btn no-data__btn btn btn-primary project-list-btn'
                                 >
-                                    {'Link new project'}
+                                    {'Link New Project'}
                                 </button>
                             </div>
                         </> :
                         <EmptyState
                             title='No project linked'
                             subTitle={{text: 'You can link a project by clicking the below button.'}}
```
Comment (sooraj-shukla) [inline webapp/src/containers/Rhs/projectList/index.tsx:112]: I think we should use contants for these.

File: webapp/src/styles/_components.scss [hunk 28]
```diff
@@ -1,5 +1,5 @@
 .height-rhs {
-    height: 100%;
+    height: calc(100% - $rhs-bottom-button-height); // 100% is the entire height and 150px is the space occupied by the button fixed at the bottom
```
Comment (sooraj-shukla) [inline webapp/src/styles/_components.scss:2]:
```suggestion
    height: calc(100% - #{$rhs-bottom-button-height}); // 100% is the entire height and 150px is the space occupied by the button fixed at the bottom
```
I think `calc` won't work this way - `calc(100% - $rhs-bottom-button-height)`.

File: webapp/src/utils/filterData.ts [hunk 29]
```diff
@@ -19,4 +19,9 @@ export const getProjectList = (data: ProjectDetails[]) => {
...
+export const getCurrentChannelSubscriptions = (data: SubscriptionDetails[], channelID: string) => (data || []).filter(((subscription) => subscription.channelID === channelID));
```
Comment (sooraj-shukla) [inline webapp/src/utils/filterData.ts:22]: `data` is required and `typeof data is array` then why do we need `(data || []).filter...`? Can we not use `data.filter...`

File: webapp/src/utils/filterData.ts [hunk 30]
```diff
@@ -19,4 +19,9 @@ export const getProjectList = (data: ProjectDetails[]) => {
...
+
+export const getCurrentChannelName = (data: ChannelList[], channelID: string) => {
+    const currentChannel = (data || []).filter(((channel) => channel.id === channelID));
+    return currentChannel[0]?.display_name;
+};
```
Comment (sooraj-shukla) [inline webapp/src/utils/filterData.ts:27]:
```suggestion
export const getCurrentChannelName = (data: ChannelList[], channelID: string) => data.find((channel) => channel.id === channelID)?.display_name;
```

It was fixed by @ayusht2810, I think it is not in sync with that branch.

**Review by sooraj-shukla — CHANGES_REQUESTED** [review]

File: webapp/src/components/toggleSwitch/index.tsx (hunk 10 above)
Comment (avas27JTG) [reply webapp/src/components/toggleSwitch/index.tsx:18]: we can do that but we should take leverage of TS wherever we can use type and literals we can get rid of import/exports

File: webapp/src/components/toggleSwitch/index.tsx (hunk 11 above)
Comment (avas27JTG) [reply webapp/src/components/toggleSwitch/index.tsx:20]: same is the case with type if a type is fixed then it's not going to change, we must utilise TS

File: webapp/src/styles/_utils.scss (hunk 13 above)
Comment (avas27JTG) [reply webapp/src/styles/_utils.scss:53]: I agree but as I mentioned we have already defined and used our utils at many places in the project so, it would be better to continue with that for now otherwise it will take time finding utils then using them here

File: webapp/src/components/loader/linear/index.tsx (hunk 19 above)
Comment (avas27JTG) [reply webapp/src/components/loader/linear/index.tsx:6]: yes but that doesn't affect anything in the UI and using `? : ' '` will be more expensive code-wise

File: webapp/src/utils/filterData.ts (hunk 29 above)
Comment (avas27JTG) [reply webapp/src/utils/filterData.ts:23]: this is a change of @ayusht2810 can you look into this and try replacing this with the one which you have fixed or is more recent in any branch

File: webapp/src/utils/filterData.ts (hunk 30 above)
Comment (avas27JTG) [reply webapp/src/utils/filterData.ts:28]: this is a change of @ayusht2810 can you look into this and try replacing this with the one which you have fixed or is more recent in any branch

File: webapp/src/containers/Rhs/projectList/index.tsx (hunk 27 above)
Comment (avas27JTG) [reply webapp/src/containers/Rhs/projectList/index.tsx:112]: yes, we can refactor texts later as for now we have just used them like this everywhere

File: webapp/src/containers/Rhs/projectDetails/index.tsx (hunk 23 above)
Comment (avas27JTG) [reply webapp/src/containers/Rhs/projectDetails/index.tsx:179]: I think it specifies because top-0 means o from top similar to margin-0

what you think for this?

File: webapp/src/containers/Rhs/projectDetails/index.tsx (hunk 2 above)
Comment (avas27JTG) [reply webapp/src/containers/Rhs/projectDetails/index.tsx:147]: this is fine for now in current branch, please update `getCurrentChannelSubscriptions` here

File: webapp/src/components/toggleSwitch/index.tsx (hunk 10 above)
Comment (sooraj-shukla) [reply webapp/src/components/toggleSwitch/index.tsx:18]: Yeah, I understand that. But for providing values, we can't use `type`, right? For that, we'll either have to use `enum` or `constant`, right?

File: webapp/src/components/toggleSwitch/index.tsx (hunk 11 above)
Comment (sooraj-shukla) [reply webapp/src/components/toggleSwitch/index.tsx:20]: `type` can be `left | right`, but by default, we want `left` here.

File: webapp/src/styles/_utils.scss (hunk 13 above)
Comment (sooraj-shukla) [reply webapp/src/styles/_utils.scss:53]: We can take the help of vscode's global-replace - `Ctrl + shift + H` to easily change it.

File: webapp/src/components/loader/linear/index.tsx (hunk 19 above)
Comment (sooraj-shukla) [reply webapp/src/components/loader/linear/index.tsx:6]: We can initialize `extraClass = ''` in the starting to overcome both these issues.

File: webapp/src/components/toggleSwitch/index.tsx (hunk 10 above)
Comment (avas27JTG) [reply webapp/src/components/toggleSwitch/index.tsx:18]: @sooraj-shukla let's sync on this I think there is come confusion

File: webapp/src/containers/Rhs/projectDetails/index.tsx (hunk 23 above)
Comment (sooraj-shukla) [reply webapp/src/containers/Rhs/projectDetails/index.tsx:179]: okay, seems fine now. I didn't understand why `top-0` was used.

File: webapp/src/containers/Rhs/projectDetails/index.tsx (hunk 22 above)
Comment (avas27JTG) [reply webapp/src/containers/Rhs/projectDetails/index.tsx:132]: nice catch

File: webapp/src/containers/Rhs/projectList/index.tsx [hunk 31]
```diff
@@ -62,11 +65,12 @@ const ProjectList = () => {
     });
 
     const {data, isSuccess, isLoading} = getApiState(plugin_constants.pluginApiServiceConfigs.getAllLinkedProjectsList.apiServiceName);
-    const projectsList = data as ProjectDetails[];
+    const projectsList = data as ProjectDetails[] ?? [];
+    const sortedProjectList = useMemo(() => [...projectsList].sort(sortProjectList), [projectsList]);
```
Comment (sooraj-shukla) [inline webapp/src/containers/Rhs/projectList/index.tsx:69]: This will still be recreated on each render since the dependency will change again and again. And hence will force the useEffect(whose dependency is `sortedProjectList`) to run on each render.

We can add a to-do comment to memoize usePluginAPI's methods and then memoize these kinds of `object type` values(which we use in `useEffect's dependency`) or we use `primitive` values in the dependency.

**Review by sooraj-shukla — APPROVED** [review]

File: README.md [hunk 32]
```diff
@@ -21,7 +21,7 @@ This plugin integrates the services of Azure DevOps in Mattermost. For a stable
 This plugin contains the following features:
 - Right-hand sidebar (RHS) shows the list of linked projects and subscriptions created for each project with separate buttons to connect an account, link or unlink a project and add or delete a subscription.
 
-- Preview work item: A preview of the work item will be created when a work item URL is posted in a channel.
+- Preview work item: A preview of the work item will be created when a work item URL is posted in a channel and the user account is conected.
```
Comment (ayusht2810) [inline README.md:24]:
```suggestion
- Preview work item: A preview of the work item will be created when a work item URL is posted in a channel and the user account is connected.
```

File: docs/developer_docs.md [hunk 33]
```diff
@@ -22,7 +22,7 @@ This plugin integrates the services of Azure DevOps in Mattermost. For a stable
 This plugin contains the following features:
 - Right-hand sidebar (RHS) shows the list of linked projects and subscriptions created for each project with separate buttons to connect an account, link or unlink a project and add or delete a subscription.
 
-- Preview work item: A preview of the work item will be created when a work item URL is posted in a channel.
+- Preview work item: A preview of the work item will be created when a work item URL is posted in a channel and the user account is conected.
```
Comment (ayusht2810) [inline docs/developer_docs.md:25]:
```suggestion
- Preview work item: A preview of the work item will be created when a work item URL is posted in a channel and the user account is connected.
```

**Review by ayusht2810 — APPROVED** [review]

File: webapp/src/plugin_constants/common.ts [hunk 34]
```diff
@@ -6,9 +6,12 @@ export const RightSidebarHeader = 'Azure DevOps';
...
-    Right = 'right',
-}
+export const boardsEventTypeMap: Record<EventType, string> = {
+    create: 'Work Item Created',
+    update: 'Work Item Updated',
+    delete: 'Work Item Deleted',
+};
 
-export const ToggleLabel = 'Show all subscriptions';
+export const channelType = {
```
Comment (manojmalik20) [inline webapp/src/plugin_constants/common.ts:15]: This is not needed. There's already a constant for this in the `mattermost-redux` package.

File: webapp/src/utils/index.ts [hunk 35]
```diff
@@ -58,14 +58,16 @@ export const getCreateTaskModalCommandArgs = (arr: Array<string>): CreateTaskCom
     };
 };
 
-export const onPressingEnterKey = (event: React.KeyboardEvent<HTMLSpanElement>, func: () => void) => {
+export const onPressingEnterKey = (event: React.KeyboardEvent<HTMLSpanElement> | React.KeyboardEvent<SVGSVGElement>, func: () => void) => {
```
Comment (manojmalik20) [inline webapp/src/utils/index.ts:61]:
```suggestion
export const onPressingEnterKey = (event: React.KeyboardEvent<HTMLSpanElement> | React.KeyboardEvent<SVGElement>, func: () => void) => {
```
Is the type for element here contains "SVG" two times?

File: webapp/src/websocket/index.ts [hunk 36]
```diff
@@ -14,3 +14,9 @@ export function handleDisconnect(store: Store<GlobalState, Action<Record<string,
         store.dispatch(toggleIsConnected(false) as Action);
     };
 }
+
+export function handleSubscriptionDeleted(store: Store<GlobalState, Action<Record<string, unknown>>>) {
+    return (_: WebsocketEventParams) => {
```
Comment (manojmalik20) [inline webapp/src/websocket/index.ts:19]: When this websocket event is received here, are you refetching the subscriptions from the API every time? You should refetch only when the RHS is open and it contains the Azure Devops plugin.
See this: https://github.com/Brightscout/mattermost-plugin-servicenow/blob/9dc472a83c62dd840e462fb7743bdf88efd9577c/webapp/src/websocket/index.ts#L56

File: webapp/src/components/card/base/styles.scss [hunk 37]
```diff
@@ -1,7 +1,7 @@
 .wrapper {
     padding: 15px;
-    border-radius: 8px;
+    border-radius: 5px;
     margin-bottom: 15px;
-    border: 1px solid rgba(var(--center-channel-color-rgb), 0.08);
+    border: 1px solid rgba(63, 67, 80, 0.16);
```
[the remark this answers is not in the document]
Comment (avas27JTG) [reply webapp/src/components/card/base/styles.scss:5]: we need to match with MM centric border color like border around slack attachment, also this remains same on every theme so didn't use variable

File: webapp/src/containers/modals/SubscribeModal/index.tsx [hunk 38]
```diff
@@ -216,7 +217,7 @@ const SubscribeModal = () => {
         >
             <>
                 {
-                    (getChannelState().isLoading || getOrganizationAndProjectState().isLoading) && <CircularLoader/>
+                    (isChannelListLoading || isOrganizationAndProjectListLoading) && <CircularLoader/>
                 }
```
[the remark this answers is not in the document]
Comment (avas27JTG) [reply webapp/src/containers/modals/SubscribeModal/index.tsx:221]: this line is modified in later PR

File: webapp/src/utils/index.ts (hunk 35 above)
Comment (avas27JTG) [reply webapp/src/utils/index.ts:61]: yes, I also noticed this but this is expected
Screenshot from 2022-09-08 15-54-04

File: webapp/src/websocket/index.ts (hunk 36 above)
Comment (avas27JTG) [reply webapp/src/websocket/index.ts:19]: even when websocket is fired we are fetching this only when `ProjectDetails` component is mounted as the API call is triggered from this component only

File: webapp/src/components/card/base/styles.scss (hunk 37 above)
[the remark this answers is not in the document]
Comment (manojmalik20) [reply webapp/src/components/card/base/styles.scss:5]: But MM may change their theme colors in the future. I think we should use variable.

File: webapp/src/websocket/index.ts (hunk 36 above)
Comment (manojmalik20) [reply webapp/src/websocket/index.ts:19]: Suppose a subscription is deleted and a websocket event was fired but at that time, the RHS was not open. So, now when we open the RHS, make sure the API doesn't get called twice. This was happening in ServiceNow. Here, the code structure is different, but still we should test it.

File: webapp/src/components/card/base/styles.scss (hunk 37 above)
[the remark this answers is not in the document]
Comment (avas27JTG) [reply webapp/src/components/card/base/styles.scss:5]: actually, MM itself is not using any variable so, we don't have any variable defined for this shade and we just copied the value from MM

image

File: docs/developer_docs.md [hunk 39]
```diff
@@ -22,7 +22,7 @@ This plugin integrates the services of Azure DevOps in Mattermost. For a stable
 This plugin contains the following features:
 - Right-hand sidebar (RHS) shows the list of linked projects and subscriptions created for each project with separate buttons to connect an account, link or unlink a project and add or delete a subscription.
 
-- Preview work item: A preview of the work item will be created when a work item URL is posted in a channel and the user account is connected.
+- Preview work item: A preview of the work item will be created when a work item URL is posted in a channel and the user account is connected to his Azure DevOps account.
```
Comment (manojmalik20) [inline docs/developer_docs.md:25]:
```suggestion
- Preview work item: A preview of the work item will be created when a work item URL is posted in a channel and the user is connected to his Azure DevOps account.
```
I think this will be better.

File: webapp/src/websocket/index.ts (hunk 36 above)
Comment (avas27JTG) [reply webapp/src/websocket/index.ts:19]: tested, and fixed this
